### ISSUE TITLE
logistic function inconsistent negative outputs

### ISSUE THREAD

**Issue by supakjk** [clab/dynet#1090]
dynet.logistic is supposed to return values in (0,1). Because of numeric issues, [0,1] should be ok. However, it seems that it sometimes return tiny negative values when the input values are large negative values.

I am using Python 2.7.14, dynet 2.0.1, and Eigen 3.3.4. The following code shows weird output:

```
import dynet as dy
import numpy as np

a = [0.0 for x in xrange(30)]
a += [-17.0502529144, -17.0516262054, -17.0403594971, -17.0335273743, -17.0677127838]
a = dy.inputTensor(a)

b = dy.logistic(a)
c = 0.5 + 0.5*dy.tanh(a*0.5)

print b.value()[-5:]
print c.value()[-5:]
```

Outputs:
```
[-5.960464477539063e-08, -5.960464477539063e-08, 3.976178675202391e-08, 4.0034372261743556e-08, 3.86889098535903e-08]
[-5.960464477539063e-08, -5.960464477539063e-08, -5.960464477539063e-08, -5.960464477539063e-08, -5.960464477539063e-08]
```

If I change the 4th line as  `a = [0.0 for x in xrange(100)]`,
The outputs are:
```
[-5.960464477539063e-08, -5.960464477539063e-08, -5.960464477539063e-08, -5.960464477539063e-08, 3.86889098535903e-08]
[-5.960464477539063e-08, -5.960464477539063e-08, -5.960464477539063e-08, -5.960464477539063e-08, -5.960464477539063e-08]
```

So, logistic results seem not consistent. I thought the implementation of logistic function is indeed `0.5 + 0.5*tanh(input*0.5)`, but not always.

**Comment by neubig** [comment]
Confirmed. The logistic function should always be between 0 and 1, so we should fix this ASAP.

**Comment by neubig** [comment]
Wait, actually in my environment I was not able to confirm this for the logistic function itself. I got the following:

    [5.960464477539063e-08, 2.9802322387695312e-08, 2.9802322387695312e-08, 2.9802322387695312e-08, 2.9802322387695312e-08]
    [5.960464477539063e-08, 2.9802322387695312e-08, -5.960464477539063e-08, 5.960464477539063e-08, 0.0]

The second one is indeed numerically unstable, but naive implementations are. We'll need to look at this more closely.

**Comment by supakjk** [comment]
The 1st line of the output changes as I try different xrange value at `a = [0.0 for x in xrange(30)]`

The 1st line of your output is also somewhat weird. a[-5] is not the largest value compared to other values but the the 1st element of the 1st line is the largest.

Can I know which code is actually used for dynet.logistic?
BTW, `scalar_logistic_sigmoid` on  https://github.com/clab/dynet/blob/7528e9dc0a8ac8afbcbb234b6f07301301778675/dynet/simd-functors.h seems using the unstable implementation using tanh.

**Comment by supakjk** [comment]
I need to directly use dy.logistic when using other loss functions like sigmoid loss.

**Comment by pmichel31415** [comment]
I have proposed a fix on this branch: https://github.com/clab/dynet/tree/stable-logistic

Can you check that it works?

for me the output of the code posted in your first comment is

Before:

```
[dynet] random seed: 4294961156
[dynet] allocating memory: 512MB
[dynet] memory allocation done.
[5.960464477539063e-08, -5.960464477539063e-08, 3.976178675202391e-08, 4.0034372261743556e-08, 3.86889098535903e-08]
[5.960464477539063e-08, -5.960464477539063e-08, -5.960464477539063e-08, 5.960464477539063e-08, 0.0]
```

After:

```
[dynet] random seed: 2683813010
[dynet] allocating memory: 512MB
[dynet] memory allocation done.
[3.937034875889367e-08, 3.931631553655279e-08, 3.976178675202391e-08, 4.0034375814457235e-08, 3.868891340630398e-08]
[-5.960464477539063e-08, -5.960464477539063e-08, -5.960464477539063e-08, -5.960464477539063e-08, -5.960464477539063e-08]
```

**Comment by neubig** [comment]
Fixed by #1200
